### PR TITLE
3DSMax 2023 & GLTFMaterial Preview

### PR DESCRIPTION
**Pull request by pandaGaume**
This PR is about support 3DSMax2023 and add the support of the new GLTFMaterial, with the introduction of new KHR_Extensions such

- KHR_materials_clearcoat
- KHR_materials_ior
- KHR_materials_sheen
- KHR_materials_specular
- KHR_materials_transmission
- KHR_materials_volume

This is the first stable preview of the exporter based on 3DSMax 2023 official pre-release.

result
_fig 1 : An export result using all the above extensions._

Note : the extensions are only working with the new GLTFMaterial. Additional work has to be conduct to make it available from PhysicalMaterial.

### CODE REVIEW

**Comment by bghgary** [comment]
Closing and reopening to trigger CI again.

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.Texture.cs [hunk 1]
```diff
@@ -212,11 +212,12 @@ private BabylonTexture ExportSpecularTexture(IIGameMaterial materialNode, float[
...
-            {
-                return ExportTexture(texMap, babylonScene, amount, allowCube);
-            }
-            return null;
+            return texMap != null ? ExportTexture(texMap, babylonScene, amount, allowCube) : null;
+        }
+        private BabylonTexture ExportPBRTexture(IIGameMaterial materialNode, string name, BabylonScene babylonScene, float amount = 1.0f, bool allowCube = false)
+        {
+            var texMap = _getTexMap(materialNode, name);
+            return texMap != null?ExportTexture(texMap, babylonScene, amount, allowCube):null;
```
Comment (bghgary) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.Texture.cs:220]: nit
```suggestion
            return texMap != null ? ExportTexture(texMap, babylonScene, amount, allowCube) : null;
```

File: 3ds Max/Max2Babylon/OnPostBuild.bat [hunk 2]
```diff
@@ -23,6 +23,7 @@ IF %1=="Debug" GOTO OnDebug
 IF %1=="Release" GOTO OnRelease
 
 :OnDebug
+REM GOTO Close
```
Comment (bghgary) [inline 3ds Max/Max2Babylon/OnPostBuild.bat:26]: should this be here?

File: SharedProjects/BabylonExport.Entities/BabylonPBRBaseSimpleMaterial.cs [hunk 3]
```diff
@@ -85,8 +90,29 @@ public BabylonPBRBaseSimpleMaterial(BabylonPBRBaseSimpleMaterial original) : bas
...
+        // SPECULAR
+        public float? metallicF0Factor { get; set; }
+        public BabylonTexture metallicReflectanceTexture { get; set; }
+        public float[] metallicReflectanceColor { get; set; } = WhiteColor();
+
+        // note this following property does NOT even exits into Babylon (vesrion 4.2.1)
+        public BabylonTexture metallicReflectanceColorTexture { get; set; }
+
+        public BabylonPBRSubSurfaceConfiguration subSurface { get; set; }
+        #endregion
```
Comment (bghgary) [inline SharedProjects/BabylonExport.Entities/BabylonPBRBaseSimpleMaterial.cs:116]: Do you need to fix this before merging?

File: SharedProjects/BabylonExport.Entities/BabylonPBRClearCoat.cs [hunk 4]
```diff
@@ -13,6 +13,8 @@ public class BabylonPBRClearCoat
 
         [DataMember]
         public float roughness { get; set; } = 0.0f;
+        [DataMember]
+        public BabylonTexture textureRoughness { get; set; }
```
Comment (bghgary) [inline SharedProjects/BabylonExport.Entities/BabylonPBRClearCoat.cs:17]: to be consistent with tintTexture
```suggestion
        public BabylonTexture roughnessTexture { get; set; }
```

File: SharedProjects/BabylonExport.Entities/BabylonPBRSheenConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        public float[] color { get; set; } 
+
+        [DataMember]
+        public BabylonTexture texture { get; set; }
+
+        [DataMember]
+        public float roughness { get; set; }
+
+        [DataMember]
+        public BabylonTexture textureRoughness { get; set; }
```
Comment (bghgary) [inline SharedProjects/BabylonExport.Entities/BabylonPBRSheenConfiguration.cs:24]: ditto

File: SharedProjects/BabylonExport.Entities/BabylonPBRSubSurfaceConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        [DataMember]
+        public float? maximumThickness { get; set; }
+
+        [DataMember]
+        public BabylonTexture thicknessTexture { get; set; }
+        [DataMember]
+        public float tintColorAtDistance { get; set; }
+        [DataMember]
+        public float[] tintColor { get; set; }
+
```
Comment (bghgary) [inline SharedProjects/BabylonExport.Entities/BabylonPBRSubSurfaceConfiguration.cs:24]: nit:
```suggestion
```

File: SharedProjects/GltfExport.Entities/KHRExtensions/KHR_materials_specular.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        public bool ShouldSerializespecularColorFactor()
+        {
+            return this.specularColorFactor != null && !this.specularColorFactor.IsAlmostEqualTo(1.0f, float.Epsilon);
+        }
+        public bool ShouldSerializespecularColorTexture()
+        {
+            return (this.specularColorTexture != null);
+        }
+    }
+
```
Comment (bghgary) [inline SharedProjects/GltfExport.Entities/KHRExtensions/KHR_materials_specular.cs:48]: nit:
```suggestion
```

File: SharedProjects/GltfExport.Entities/KHRExtensions/KHR_materials_volume.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        public bool ShouldSerializeattenuationDistance()
+        {
+            return this.attenuationDistance != null && this.attenuationDistance != 0 && this.attenuationDistance != float.PositiveInfinity;
+        }
+        public bool ShouldSerializeattenuationColor()
+        {
+            return this.attenuationColor != null && this.attenuationColor.Length == 3 && !this.attenuationColor.IsAlmostEqualTo(1.0f, float.Epsilon);
+        }
+    }
+
```
Comment (bghgary) [inline SharedProjects/GltfExport.Entities/KHRExtensions/KHR_materials_volume.cs:50]: nit:
```suggestion
```

File: SharedProjects/BabylonExport.Entities/BabylonPBRBaseSimpleMaterial.cs (hunk 3 above)
Comment (pandaGaume) [reply SharedProjects/BabylonExport.Entities/BabylonPBRBaseSimpleMaterial.cs:116]: Not really, this is part of the process by step. following the discussion you had with seb, i do not know yet what is your choice about adding gltf support to simpleMaterial or  refactor the whole material export to target the huge PbrMaterial. the second options is a bit heavy, so i might wait for the first options until the last minute :-)

File: SharedProjects/BabylonExport.Entities/BabylonPBRClearCoat.cs (hunk 4 above)
Comment (pandaGaume) [reply SharedProjects/BabylonExport.Entities/BabylonPBRClearCoat.cs:17]: This name is exact mapping of Babylon JS class, which is itself not consistent.

File: SharedProjects/BabylonExport.Entities/BabylonPBRSheenConfiguration.cs (hunk 5 above)
Comment (pandaGaume) [reply SharedProjects/BabylonExport.Entities/BabylonPBRSheenConfiguration.cs:24]: ditto also :-)

File: 3ds Max/Max2Babylon/OnPostBuild.bat (hunk 2 above)
Comment (pandaGaume) [reply 3ds Max/Max2Babylon/OnPostBuild.bat:26]: no...

Comment (pandaGaume) [reply 3ds Max/Max2Babylon/OnPostBuild.bat:26]: no

File: SharedProjects/BabylonExport.Entities/BabylonPBRBaseSimpleMaterial.cs (hunk 3 above)
Comment (pandaGaume) [reply SharedProjects/BabylonExport.Entities/BabylonPBRBaseSimpleMaterial.cs]: Removed by last commit to targeting BabylonPBRMaterial

File: SharedProjects/BabylonExport.Entities/BabylonPBRMaterial.cs [hunk 9]
```diff
@@ -220,9 +237,9 @@ public BabylonPBRMaterial(BabylonPBRBaseSimpleMaterial origin) : base(origin.id)
             usePhysicalLightFalloff = true;
             useEmissiveAsIllumination = true;
 
-             ambient = new[] { 0f, 0f, 0f };
-            reflectivity = new[] { 1f, 1f, 1f };
-            reflection = new[] { 1f, 1f, 1f };
+             ambient = BlackColor();
```
Comment (bghgary) [inline SharedProjects/BabylonExport.Entities/BabylonPBRMaterial.cs:240]: nit:
```suggestion
            ambient = BlackColor();
```

**Review by bghgary — APPROVED** [review]
Seems fine. Some nits still left but they don't have to be fixed if you don't want to.

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.Material.GLTF.cs [hunk 10]
```diff
@@ -517,5 +529,30 @@ internal void ExportSpecular(GltfMaterialDecorator maxDecorator, BabylonPBRMater
 
             }
         }
+
+        private void ExportAlphaMode(GltfMaterialDecorator maxDecorator, BabylonPBRMaterial target)
+        {
+            var a = maxDecorator.AlphaMode;
+            switch (a)
```
Comment (bghgary) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.Material.GLTF.cs:536]: nit:
```suggestion
            switch (maxDecorator.AlphaMode)
```

File: 3ds Max/Max2Babylon/Exporter/BabylonExporter.Material.GLTF.cs [hunk 11]
```diff
@@ -238,6 +238,7 @@ internal void ExportPhysicalBase(GltfMaterialDecorator maxDecorator, BabylonPBRM
                 babylonMaterial.metallic = maxDecorator.Metalness;
                 babylonMaterial.roughness = maxDecorator.Roughness;
                 babylonMaterial.emissive = maxDecorator.EmissionColor?.ToArray();
+                babylonMaterial.twoSidedLighting = maxDecorator.DoubleSided;
```
Comment (bghgary) [inline 3ds Max/Max2Babylon/Exporter/BabylonExporter.Material.GLTF.cs:241]: I'm not sure Max's doubleSided means twoSidedLighting. glTF's double sided means `backfaceCulling = false` and `twoSidedLighting = true`. I'm not sure what Max's doubleSided means.

**Pull request merged**